### PR TITLE
Android retry

### PR DESCRIPTION
**Pull request by Sarkar**
Refactor retry to network requests instead of at the Branch level. Similar logic to @shortstuffsushi PR. 

@qinweigong - since you have a big PR request coming in, let's figure out the best strategy to avoid conflicts yeah?

### CODE REVIEW

File: Branch-SDK/src/io/branch/referral/BranchConfig.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package io.branch.referral;
+
+/*
+ *
+ */
+public class BranchConfig {
```
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/BranchConfig.java]: Do you have plan to move more stuff into this class? If only for this TAG, maybe we can just put it into PrefHelper.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 2]
```diff
@@ -26,128 +27,183 @@
...
-		}
-	}
+    public static final String NO_TAG_VALUE = "no_tag";
+    public static final int NO_CONNECTIVITY_STATUS = -1009;
+    public static final int NO_API_KEY_STATUS = -1234;
+
+    private static final String SDK_VERSION = "1.4.4";
+    private static final int DEFAULT_TIMEOUT = 3000;
+
+    protected PrefHelper prefHelper_;
```
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: Besides moving prefHelper_ here, can I assume all the other lines are for tab -> spaces changes? It's hard to compare so many lines one by one :-)

**Comment by shortstuffsushi** [comment]
Adding to @qinweigong's statement, it would be good in the future to separate formatting from pull request content, or at least make a separate commit. Definitely makes diffing harder.

**Comment by Sarkar** [comment]
Sorry about that guys. I should've done that. I can say that the biggest
diff is inside the base GET and POST functions adding logic to retry while
having a retry count. Essentially what Graham wrote for iOS  perhaps
looking at individual commits will help

On Thursday, April 9, 2015, Graham Mueller <notifications@github.com> wrote:

> Adding to @qinweigong <https://github.com/qinweigong>'s statement, it
> would be good in the future to separate formatting from pull request
> content, or at least make a separate commit. Definitely makes diffing
> harder.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BranchMetrics/Branch-Android-SDK/pull/58#issuecomment-91418246>
> .
>


-- 

*Sahil Verma*

*Branch Metrics *

w branch.io | c 510-386-1750 | s sahilroyverma

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 3]
```diff
@@ -26,128 +27,183 @@
...
+            return new ServerResponse(tag, NO_CONNECTIVITY_STATUS);
+        } catch (Exception ex) {
+            if (log) {
+                PrefHelper.Debug(getClass().getSimpleName(), "Exception: " + ex.getMessage());
+            }
+            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.HONEYCOMB) {
+                if (ex instanceof NetworkOnMainThreadException)
+                    Log.e(BranchConfig.TAG, "Branch Error: Don't call our synchronous methods on the main thread!");
+            }
+            return new ServerResponse(tag, 500);
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: 500 doesn't seem like the right code here -- indicates server failure.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 4]
```diff
@@ -26,128 +27,183 @@
...
+            request.setHeader("Content-type", "application/json");
+            HttpResponse response = getGenericHttpClient(timeout).execute(request);
+            int statusCode = response.getStatusLine().getStatusCode();
+            boolean isRetryableStatusCode = statusCode >= 500;
+
+            if ((retryNumber < prefHelper_.getRetryCount()) && isRetryableStatusCode) {
+                retryNumber +=1;
+                make_restful_post(body, url, tag, timeout, retryNumber, log, linkData);
+            } else {
+                return processEntityForJSON(response.getEntity(), response.getStatusLine().getStatusCode(), tag, log, null);
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: Seems this else and the next line could be combined.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 5]
```diff
@@ -26,128 +27,183 @@
...
+            }
+
+            body.put("sdk", "android" + SDK_VERSION);
+            if (log) {
+                PrefHelper.Debug(BranchConfig.TAG, "posting to " + url);
+                PrefHelper.Debug(BranchConfig.TAG, "Post value = " + body.toString(4));
+            }
+            HttpPost request = new HttpPost(url);
+            request.setEntity(new ByteArrayEntity(body.toString().getBytes("UTF8")));
+            request.setHeader("Content-type", "application/json");
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: Content-Type should be PascalCased, though according to the RFC, the server should accept it either way.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 6]
```diff
@@ -26,128 +27,183 @@
...
+    }
+
+    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout, boolean log) {
+        return make_restful_post(body, url, tag, timeout, 0, log, null);
+    }
+
+    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout,
+                                            int retryNumber, boolean log, BranchLinkData linkData) {
+        try {
+            if (body.has("app_id") && body.getString("app_id").equals(PrefHelper.NO_STRING_VALUE)) {
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: On iOS, I'm adding the app_id myself in this method, rather than having it passed into this method. That way, rather than having to remember this for every request, things that we *always* want are added to each request. I also add things like sdk=iOS0.x.x, etc.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 7]
```diff
@@ -26,128 +27,183 @@
...
+        } catch (IOException ex) {
+            if (log) {
+                PrefHelper.Debug(getClass().getSimpleName(), "IO exception: "
+                        + ex.getMessage());
+            }
+        }
+        return result;
+    }
+
+    //region GET Requests
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: Unrelated to anything, but does IntelliJ respect regions?

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java (hunk 6 above)
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: @shortstuffsushi I am make that change in my PR, FYI.

File: Branch-SDK/src/io/branch/referral/BranchConfig.java (hunk 1 above)
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/BranchConfig.java]: It seems like this would be an appropriate place for things like the release number -- it's used for that in iOS. It also sets the base server url depending on whether you're in debug.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java (hunk 6 above)
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: This is all super valuable feedback but it actually doesn't pertain to what
I changed. Let's sync up about this tomorrow? So I can go over with you two
what I changed then we finalize?

On Thursday, April 9, 2015, qinweigong <notifications@github.com> wrote:

> In Branch-SDK/src/io/branch/referral/RemoteInterface.java
> <https://github.com/BranchMetrics/Branch-Android-SDK/pull/58#discussion_r28121046>
> :
>
> > +    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout) {
> > +        return make_restful_post(body, url, tag, timeout, 0, true, null);
> > +    }
> > +
> > +    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout, BranchLinkData linkData) {
> > +        return make_restful_post(body, url, tag, timeout, 0, true, linkData);
> > +    }
> > +
> > +    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout, boolean log) {
> > +        return make_restful_post(body, url, tag, timeout, 0, log, null);
> > +    }
> > +
> > +    public ServerResponse make_restful_post(JSONObject body, String url, String tag, int timeout,
> > +                                            int retryNumber, boolean log, BranchLinkData linkData) {
> > +        try {
> > +            if (body.has("app_id") && body.getString("app_id").equals(PrefHelper.NO_STRING_VALUE)) {
>
> @shortstuffsushi <https://github.com/shortstuffsushi> I am make that
> change in my PR, FYI.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BranchMetrics/Branch-Android-SDK/pull/58/files#r28121046>
> .
>


-- 

*Sahil Verma*

*Branch Metrics *

w branch.io | c 510-386-1750 | s sahilroyverma

Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: Ah, see that now -- lost that cuz of the big block diff :wink: Sounds good though.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java [hunk 8]
```diff
@@ -26,128 +27,183 @@
...
+    public ServerResponse make_restful_get(String url, String tag, int timeout) {
+        return make_restful_get(url, tag, timeout, 0, true);
+    }
+
+    public ServerResponse make_restful_get(String url, String tag, int timeout, int retryNumber, boolean log) {
```
Comment (shortstuffsushi) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: My only other feedback about this change is the fact that the method w/ retry number is exposed here. On iOS, the actual retry method is private -- not sure if we care to do the same here.

File: Branch-SDK/src/io/branch/referral/BranchConfig.java (hunk 1 above)
Comment (qinweigong) [inline Branch-SDK/src/io/branch/referral/BranchConfig.java]: My bad. I created BNCConfig on iOS side, but too lazy to do the same on Android side at that time.
Anyway, I'd prefer moving all the stuff to here in one shot. People getting this release will find this class with so little content. That doesn't look good :-)

Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/BranchConfig.java]: @shortstuffsushi this is exactly what I intend for thsi class. I'd rather move final statics like the end point, URL, version number, etc all in one place so it's never manually specified else where. 

@qinweigong in that case that will be a separate PR then.

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java (hunk 2 above)
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: yep! + the retry count

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java (hunk 5 above)
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: This is a good catch. I'll add this

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java (hunk 7 above)
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: heck yes! code folding all the way

File: Branch-SDK/src/io/branch/referral/RemoteInterface.java (hunk 8 above)
Comment (Sarkar) [inline Branch-SDK/src/io/branch/referral/RemoteInterface.java]: i thought about this too. the base method should be private for both GETs and POSTs. I'll make this change as well.
